src/js/angular/guides/tour-lib-services/shepherd.service.js

```diff
@@ -0,0 +1,543 @@
+     *       <ul>
+     *           <li>
+     *               <b>clickable-element</b> - the element is clickable like link, button... Client of workbench can click on this element.
+     *           </li>
+     *           <li>
+     *               <b>read-only-element</b> - the element is readable. Client of workbench can only read this component.
+     *           </li>
+     *           <li>
+     *               <b>input-element</b> - the element is some kind ot input like text, text area ... . Client of workbench can enter data in the element.
+     *           </li>
+     *        </ul>
+     *     </li>
+     *     <li>
+     *         <b>maxWaitTime</b> - some element need more time to be present on the page. This is the max waiting time element to be visible
+     *         on the page. Default value is 3 second.
+     *     </li>
+     *     <li>
+     *         <b>canBePaused</b> - some element need user interaction with some other element on the page before appear. Step which uses such
+     *         element can be persist when guide is paused. For example if step is "Click on sub menu "Repositories" of menu "Setup""
+     *     </li>
+     *     <li>
+     *         <b>onNextClick</b> - function which will be executed when nex button is clicked.
+     *     </li>
+     *      <li>
+     *         <b>onPreviousClick</b> - function which will be executed when previous button is clicked.
+     *     </li>
+     * </ul>
+     */
+    this._addGuideSteps = (guide, stepsDescriptions) => {
+        if (stepsDescriptions.length > 0) {
+            guide.addStep(this._getFirstStep(guide, stepsDescriptions));
+        }
+        if (stepsDescriptions.length > 2) {
+            for (let index = 1; index < stepsDescriptions.length - 1; index++) {
+                guide.addStep(this._getMiddleStep(guide, stepsDescriptions, index));
+            }
+        }
+        if (stepsDescriptions.length > 1) {
+            guide.addStep(this._getLastStep(guide, stepsDescriptions));
+        }
+    }
+
+    /**
+     * Starts <code>guide</code> from <code>startStepId</code>
+     * @param guide - the guid to be started.
+     * @param startStepId - star step id. If it is not present the <code>guide</code> will be started from beginning.
+     * @private
+     */
+    this._startGuide = (guide, startStepId) => {
+        const stepIndex = this._getStepIndex(guide, startStepId);
+        const step = guide.steps[stepIndex];
+        const url = step.options.url;
+        if (!!url) {
+            $location.path(url);
+        }
+
+        GuideUtils.waitFor(step.options.attachTo.element, step.options.maxWaitTime)
+            .then(() => guide.show(stepIndex))
+            .catch(() => guide.cancel());
+    }
+
+    this._subscribeToGuideCanceled = () => {
+        if (!this.guideCancelSubscription) {
+            this.guideCancelSubscription = Shepherd.on('cancel', () => {
+                this._clearLocaleStorage();
```

Comment:
Can't you use the cancel function from this service here instead of repeating this?

Reply:
I removed duplicated code in cancel function of ShepherdService. This function will be called if active tour is canceled.